vir/src/callable_idents.rs

```diff
@@ -1,117 +1,159 @@
     }
     fn name(&self) -> &'vir str {
         self.0
     }
-    fn arity(&self) -> A {
-        self.1
+    fn arity(&self) -> &A {
+        &self.1
     }
 }
 
-#[sealed]
-pub trait Arity: Copy {
-    fn args(&self) -> &[Type];
-    fn len_matches(&self, len: usize) -> bool;
-    fn check<'vir, Curr: 'vir, Next: 'vir>(&self, name: &str, args: &[ExprGen<'vir, Curr, Next>]) {
-        if cfg!(debug_assertions) {
-            let args_len = args.len();
-            assert!(self.len_matches(args_len), "{name} called with {args_len} args (expected {})", self.args().len());
-            for (_arg, _ty) in args.iter().zip(self.args()) {
-                // TODO: check that the types match
-            }
-        }
-        // TODO: return result type
+#[derive(Debug, Clone, Copy)]
+pub struct DomainIdent<'vir, A: Arity<'vir>>(&'vir str, A);
+impl<'vir, A: Arity<'vir>> CallableIdent<'vir, A> for DomainIdent<'vir, A> {
+    fn new(name: &'vir str, args: A) -> Self {
+        Self(name, args)
+    }
+    fn name(&self) -> &'vir str {
+        self.0
+    }
+    fn arity(&self) -> &A {
+        &self.1
     }
 }
+pub type DomainIdentUnknownArity<'vir> = DomainIdent<'vir, UnknownArityAny<'vir, DomainParamData<'vir>>>;
+
+#[sealed]
+pub trait Arity<'vir>: Copy {
+    type Arg;
+    fn args(&self) -> &'vir [Self::Arg];
+    fn check_len_matches(&self, name: &str, len: usize);
+}
 #[sealed]
-impl<const N: usize> Arity for KnownArity<'_, N> {
-    fn args(&self) -> &[Type] {
+impl<'vir, T, const N: usize> Arity<'vir> for KnownArityAny<'vir, T, N> {
+    type Arg = T;
+    fn args(&self) -> &'vir [T] {
         &self.0
     }
-    fn len_matches(&self, _len: usize) -> bool {
-        true
-    }
+    fn check_len_matches(&self, _name: &str, _len: usize) {}
 }
 #[sealed]
-impl Arity for UnknownArity<'_> {
-    fn args(&self) -> &[Type] {
-        self.0
+impl<'vir, T> Arity<'vir> for UnknownArityAny<'vir, T> {
+    type Arg = T;
+    fn args(&self) -> &'vir [T] {
+        &self.0
     }
-    fn len_matches(&self, len: usize) -> bool {
-        self.0.len() == len
+    fn check_len_matches(&self, name: &str, len: usize) {
+        assert_eq!(self.0.len(), len, "{name} called with {len} args (expected {})", self.0.len());
     }
 }
 
-#[derive(Debug, Clone, Copy)]
-pub struct KnownArity<'vir, const N: usize>([Type<'vir>; N]);
-impl<'vir, const N: usize> KnownArity<'vir, N> {
-    pub const fn new(types: [Type<'vir>; N]) -> Self {
+trait CheckTypes<'vir> {
+    fn check_types<Curr: 'vir, Next: 'vir>(&self, name: &str, args: &[ExprGen<'vir, Curr, Next>]);
+}
+impl<'vir, A: Arity<'vir, Arg = Type<'vir>>> CheckTypes<'vir> for A {
+    fn check_types<Curr: 'vir, Next: 'vir>(&self, name: &str, args: &[ExprGen<'vir, Curr, Next>]) {
+        if cfg!(debug_assertions) {
+            self.check_len_matches(name, args.len());
+            for (_arg, _ty) in args.iter().zip(self.args().into_iter()) {
+                // TODO: check that the types match
```

Comment:
Is this just equality or is something more involved supposed to happen?

Reply:
It was meant to be just equality... before I switched to doing the domain generics. So now we'll need to check equality up to domain type params.